plugins/woocommerce/src/Blocks/Domain/Services/Hydration.php

```diff
@@ -38,17 +41,139 @@ public function __construct( AssetDataRegistry $asset_data_registry ) {
+		$this->cache_store_notices();
+
+		$preloaded_data = array();
+
+		if ( null !== $controller_class ) {
+			try {
+				$response = $this->get_response_from_controller( $controller_class, $path );
+				if ( $response ) {
+					$preloaded_data = array(
+						'body'    => $response->get_data(),
+						'headers' => $response->get_headers(),
+					);
+				}
+			} catch ( \Exception $e ) {
+				// This is executing in frontend of the site, a failure in hydration should not stop the site from working.
+				wc_get_logger()->warning(
+					'Error in hydrating REST API request: ' . $e->getMessage(),
+					array(
+						'source'    => 'blocks-hydration',
+						'data'      => array(
+							'path'       => $path,
+							'controller' => $controller_class,
+						),
+						'backtrace' => true,
+					)
+				);
+			}
+		} else {
+			// Preload the request and add it to the array. It will be $preloaded_requests['path']  and contain 'body' and 'headers'.
+			$preloaded_requests = rest_preload_api_request( [], $path );
+			$preloaded_data     = $preloaded_requests[ $path ] ?? [];
+		}
 
 		$this->restore_cached_store_notices();
 		$this->restore_nonce_check();
 
 		// Returns just the single preloaded request, or an empty array if it doesn't exist.
-		return $preloaded_requests[ $path ] ?? [];
+		return $preloaded_data;
+	}
+
+	/**
+	 * Helper method to generate GET response from a controller. Also fires the `rest_request_after_callbacks` for backward compatibility.
+	 *
+	 * @param string $controller_class Controller class FQN that will respond to the request.
+	 * @param string $path             Request path regex.
+	 *
+	 * @return false|mixed|null Response
+	 */
+	private function get_response_from_controller( $controller_class, $path ) {
+		if ( null === $controller_class ) {
+			return false;
+		}
+
+		$request           = new \WP_REST_Request( 'GET', $path );
+		$schema_controller = StoreApi::container()->get( SchemaController::class );
+		$controller        = new $controller_class(
+			$schema_controller,
+			$schema_controller->get( $controller_class::SCHEMA_TYPE, $controller_class::SCHEMA_VERSION )
+		);
+
+		$controller_args = is_callable( array( $controller, 'get_args' ) ) ? $controller->get_args() : [];
+
+		if ( empty( $controller_args ) ) {
+			return false;
+		}
+
+		// Get the handler that responds to read request.
+		$handler = current(
+			array_filter(
+				$controller_args,
+				function ( $method_handler ) {
+					return is_array( $method_handler ) && isset( $method_handler['methods'] ) && \WP_REST_Server::READABLE === $method_handler['methods'];
+				}
+			)
+		);
+
+		if ( ! $handler ) {
+			return false;
+		}
+
+		$response = call_user_func_array( $handler['callback'], array( $request ) );
+
+		/**
+		 * For backward compatibility with WC 8.6 and earlier, we manually call this filter that is otherwise called by WP's REST API. This provides the opportunity for 3PD plugins to hook into and change the response data.
+		 *
```

Comment:
```suggestion
		 * Note that altering the Store API final response is something we discourage and won't ultimately support. This will be deprecated in favor of better, more integrated solutions.
```